raft/server.cc

```diff
@@ -115,7 +115,13 @@ class server_impl : public rpc_server, public server {
     condition_variable _applied_index_changed;
 
     struct stop_apply_fiber{}; // exception to send when apply fiber is needs to be stopepd
-    queue<std::variant<std::vector<log_entry_ptr>, snapshot_descriptor>> _apply_entries = queue<std::variant<std::vector<log_entry_ptr>, snapshot_descriptor>>(10);
+
+    struct removed_from_config{}; // sent to applier_fiber when we're not a leader and we're outside the current configuration
```

Comment:
Can we just move the check of  `_current_rpc_config.contains(_id)` to io fiber instead of introducing a new message. We never suppose to add nodes with the same id again. (But we probably do in the tests unfortunately).

Reply:
Then you would still have the race that this PR is fixing, just at a different moment.
1. `io_fiber` sends the `batch.committed` entries
2. `io_fiber` updates `_current_rpc_config`
3. `applier_fiber` performs this check before fetching the message from step 1, drops the waiters.

Reply:
You mean there may be more than one batch in the queue and applier will drop waiters before processing all of them? Yes, this probably can happen.

Reply:
@gleb-cloudius it doesn't even take an extra batch, applier queue can contain snapshot descriptors as well. So just one batch and one snapshot descriptor.